feat: Introduce memory usage debugging for LengthBasedSplitter

erdncyz opened the pull request:
Problem: "The frame is bigger than expected" error with 808MB+ data packets from iOS devices.
Solution: Increased MAX_FRAME_SIZE from 1MB-20MB to 1GB across all services.
Changes Made:
Updated 6 files: webinspector, afc, usbmux, plist-service, notification-proxy, house-arrest
Fixed TypeScript compilation errors
Version bumped to 2.8.5
Result: Now handles large iOS device data frames (up to 1GB) without errors.
Repository: Changes pushed to https://github.com/erdncyz/appium-ios-device

KazuCocoa commented on package.json:
```diff
@@ -4,7 +4,7 @@
     "appium"
   ],
-  "version": "2.8.4",
+  "version": "2.8.5",
```
This version increment occurs through our auto-release. Please do not modify here

erdncyz replied:
ı done @KazuCocoa

KazuCocoa commented on lib/webinspector/index.js:
```diff
@@ -88,7 +88,7 @@ class WebInspectorService extends BaseServiceSocket {
     this._majorOsVersion = majorOsVersion;
 
-    if (!isSimulator && !_.isNil(majorOsVersion) && majorOsVersion < PARTIAL_MESSAGE_SUPPORT_DEPRECATION_VERSION) {
+    if (!isSimulator && !_.isNil(majorOsVersion) && majorOsVersion !== undefined && majorOsVersion < PARTIAL_MESSAGE_SUPPORT_DEPRECATION_VERSION) {
```
`_.isNil` inclues nil/undefined check
https://lodash.com/docs#isNil

erdncyz commented:
Hi @mykola-mokhnach , Thank you for the excellent feedback! You're absolutely right.

Instead of blindly increasing limits, I've implemented:

1. **Intelligent debugging system** with configurable logging
2. **Service identification** to pinpoint which service causes large frames
3. **Memory monitoring** to prevent OOM issues
4. **Comprehensive debugging guide** for investigation

Now users can:
- Enable debug logging to identify the root cause
- See exactly which service and data size is problematic
- Monitor memory usage during large frame events
- Get actionable debugging steps

This approach addresses your concerns about OOM risks and provides tools to investigate the underlying issue rather than just masking it with larger limits.

mykola-mokhnach commented on lib/util/transformer/length-based-splitter.js:
```diff
@@ -48,8 +58,37 @@ export class LengthBasedSplitter extends Stream.Transform {
+      // Log memory usage if monitoring is enabled
+      if (DEBUG_CONFIG.enableMemoryMonitoring) {
+        const memUsage = process.memoryUsage();
+        log.warn(`Memory usage - RSS: ${(memUsage.rss / (1024 * 1024)).toFixed(2)} MB, Heap: ${(memUsage.heapUsed / (1024 * 1024)).toFixed(2)} MB`);
```
consider using the [toReadableSizeString](https://github.com/appium/appium/blob/df56c667e8b1d144ba0310a9567581641c0d4347/packages/support/lib/util.js#L238) helper

mykola-mokhnach commented on DEBUG_FRAME_SIZE.md:
```diff
@@ -0,0 +1,107 @@
+# Debugging Large Frame Size Issues
```
👍

erdncyz replied:
Hi @mykola-mokhnach 

I did

mykola-mokhnach commented on DEBUG_FRAME_SIZE.md:
````diff
@@ -0,0 +1,107 @@
+### Option 1: Increase Limits (Not Recommended)
+```javascript
+// Only use this after investigation
+const MAX_FRAME_SIZE = 1024 * 1024 * 1024; // 1GB
````
this is not very practical as such change would need to recompile the whole module. We may also have this limit configurable as environment variable

erdncyz commented:
Hi @mykola-mokhnach 

I did all of your request. Can you check again please

mykola-mokhnach commented on docs/DEBUG_FRAME_SIZE.md:
````diff
@@ -0,0 +1,113 @@
+```javascript
+// Disable detailed debugging if WebInspector is the issue
+const capabilities = {
+  'appium:webviewDevtoolsPort': -1,
````
I'm not sure about these capabilities. Where are they documented?

mykola-mokhnach replied:
Sorry, I don't observe this comment being resolved

erdncyz replied:
You're absolutely right! I removed the undocumented `appium:webviewDevtoolsPort` capability and kept only the documented `appium:enableWebviewDetailsCollection`. 

The documentation now suggests users check their specific Appium driver documentation for available capabilities rather than providing potentially incorrect examples. Thanks for catching that!

mykola-mokhnach commented on lib/utilities.js:
```diff
@@ -4,6 +4,24 @@ import _ from 'lodash';
+ * @param {number} [decimals=2] - Number of decimal places
+ * @returns {string} Human-readable size string (e.g., "1.23 MB", "456.78 KB")
+ */
+function toReadableSizeString(bytes, decimals = 2) {
```
I don't see a need to reimplement this method simply import it as is from the support package 
```
import {util } from '@appium/support';

....

console.log(util.toReadableSizeString(sizeBytes));
...
```

erdncyz commented:
hi @mykola-mokhnach ,
ı did all of your comments. Thank you very much

mykola-mokhnach commented on lib/util/transformer/length-based-splitter.js:
```diff
@@ -96,6 +118,39 @@ export class LengthBasedSplitter extends Stream.Transform {
+    log.error(`Service: ${this.serviceName}, Frame offset: ${this.lengthFieldOffset}, Frame length: ${this.lengthFieldLength}`);
+    
+    // Suggest debugging steps
+    log.error('To debug this issue, set environment variables:');
```
I assume these log messages should not be shown if corresponding environment variables have been already activated

mykola-mokhnach commented on lib/util/transformer/length-based-splitter.js:
```diff
@@ -96,6 +118,39 @@ export class LengthBasedSplitter extends Stream.Transform {
+    log.error('  APPIUM_IOS_LOG_LARGE_FRAMES=true');
+    log.error('  APPIUM_IOS_MONITOR_MEMORY=true');
+    
+    throw new Error(errorMsg);
```
it would be easier for the compiler to understand the code flow if this function returns the error object instead of throwing it, e.g. replace:

```
this._handleFrameSizeError(messageLength);
```

with 

```
throw this._makeFrameSizeError(messageLength);
```

and change the last expression in the method to 

```
return new Error(errorMsg);
```

mykola-mokhnach commented on lib/utilities.js:
```diff
@@ -3,6 +3,9 @@ import { upgradeToSSL, enableSSLHandshakeOnly } from './ssl-helper';
 import _ from 'lodash';
 import { log } from './logger';
 import { findDeveloperImage } from './imagemounter/utils/list_developer_image';
+import { util } from '@appium/support';
```
is this import used?

erdncyz commented:
hi @mykola-mokhnach ,
ı did all of your comments. Thank you very much

mykola-mokhnach commented on lib/house-arrest/index.js:
```diff
@@ -18,10 +19,11 @@ class HouseArrestService extends BaseServiceSocket {
       lengthFieldOffset: 0,
       lengthFieldLength: 4,
       lengthAdjustment: 4,
+      serviceName: 'HouseArrest',
```
nit: you could also use `this.constructor.name` to avoid hardcoding service names

mykola-mokhnach commented on lib/util/transformer/length-based-splitter.js:
```diff
@@ -96,6 +118,42 @@ export class LengthBasedSplitter extends Stream.Transform {
+    
+    // Only suggest debugging steps if debug mode is not already enabled
+    if (!DEBUG_CONFIG.enableFrameSizeLogging) {
+      log.error('To debug this issue, set environment variables:');
```
you may also simply provide the link to the above document here instead of listing env variable names

mykola-mokhnach reviewed (approved):
LGTM with minor comments

Please also sign the CLA, so we could merge the PR

mykola-mokhnach commented on lib/house-arrest/index.js:
```diff
@@ -78,6 +79,8 @@ class HouseArrestService extends BaseServiceSocket {
       setTimeout(() => reject(new Error(`Failed to receive any data within the timeout: ${timeout}`)), timeout);
     });
   }
+
```
this space is unncessary

mykola-mokhnach commented on lib/util/transformer/length-based-splitter.js:
```diff
@@ -80,6 +101,12 @@ export class LengthBasedSplitter extends Stream.Transform {
     this._frameBufferIndex = 0;
   }
 
+
```
this space is unnecessary

mykola-mokhnach commented on lib/base-service.js:
```diff
@@ -21,6 +21,8 @@ class BaseServiceSocket {
       this._socketClient.end();
     }
   }
+
```
same as above

mykola-mokhnach commented on lib/utilities.js:
```diff
@@ -4,6 +4,8 @@ import _ from 'lodash';
 import { log } from './logger';
 import { findDeveloperImage } from './imagemounter/utils/list_developer_image';
 
+
```
same as above

erdncyz commented:
hi again @mykola-mokhnach,

ı did all of your comments. Thank youuu

mykola-mokhnach commented:
Unfortunately I don't observe CLA being [signed](https://github.com/appium/appium-ios-device/pull/202#issuecomment-3143624673). Could you please make sure your github account email matches to the one used to sign the CLA?

mykola-mokhnach commented on docs/DEBUG_FRAME_SIZE.md:
```diff
@@ -0,0 +1,116 @@
+// Disable detailed debugging if WebInspector is the issue
+const capabilities = {
+  // Disable WebInspector debugging to reduce data transfer
+  'appium:enableWebviewDetailsCollection': false,
```
as mentioned earlier this capability is not applicable for iOS drivers

mykola-mokhnach commented on README.md:
```diff
@@ -114,6 +114,10 @@ async function checkAndMountDeveloperImage(udid) {
 
 * If exists, `USBMUXD_SOCKET_ADDRESS` is used to get usbmuxd socket address. Mostly useful in cases where the `usbmuxd` is run by a non-root user.
 
+## Documentation
```
Please rename to `Troubleshooting`

erdncyz reviewed (commented):
ı did everything comments

erdncyz commented:
Hi @mykola-mokhnach ,

ı try many methods but ı didn't solve this CLA issue. Can you help me please. ı added all my email account.

KazuCocoa reviewed (commented):
Left a few comments, otherwise lgtm entirely 👍

KazuCocoa commented on lib/util/transformer/length-based-splitter.js:
```diff
@@ -147,9 +147,9 @@ export class LengthBasedSplitter extends Stream.Transform {
-    if (!DEBUG_CONFIG.enableFrameSizeLogging) {
-      log.error('To debug this issue, see: https://github.com/erdncyz/appium-ios-device/blob/master/docs/DEBUG_FRAME_SIZE.md');
-    }
+                    if (!DEBUG_CONFIG.enableFrameSizeLogging) {
```
Do not need to modify the indentiations in this commit

KazuCocoa commented on lib/utilities.js:
```diff
@@ -253,8 +253,24 @@ async function fetchImageFromGithubRepo(udid, opts) {
+function getMaxFrameLength(defaultSize) {
+  const envValue = process.env.APPIUM_IOS_MAX_FRAME_SIZE;
+  if (envValue) {
+          const parsed = _.parseInt(envValue, 10);
```
```suggestion
    const parsed = _.parseInt(envValue, 10);
```

KazuCocoa commented on lib/webinspector/index.js:
```diff
@@ -69,13 +70,13 @@ class WebInspectorService extends BaseServiceSocket {
 
     // set the largest fragment size for the socket, if the option is there
-    if (_.isFunction(socketClient.setMaxSendFragment) && !_.isNil(socketChunkSize) && socketChunkSize > 0) {
+            if (_.isFunction(socketClient.setMaxSendFragment) && _.isNumber(socketChunkSize) && socketChunkSize > 0) {
```
```suggestion
    if (_.isFunction(socketClient.setMaxSendFragment) && _.isNumber(socketChunkSize) && socketChunkSize > 0) {
```

KazuCocoa commented on lib/webinspector/index.js:
```diff
@@ -88,7 +89,7 @@ class WebInspectorService extends BaseServiceSocket {
     this._majorOsVersion = majorOsVersion;
 
-    if (!isSimulator && !_.isNil(majorOsVersion) && majorOsVersion < PARTIAL_MESSAGE_SUPPORT_DEPRECATION_VERSION) {
+        if (!isSimulator && !_.isNil(majorOsVersion) && majorOsVersion < PARTIAL_MESSAGE_SUPPORT_DEPRECATION_VERSION) {
```
```suggestion
    if (!isSimulator && !_.isNil(majorOsVersion) && majorOsVersion < PARTIAL_MESSAGE_SUPPORT_DEPRECATION_VERSION) {
```

KazuCocoa reviewed (commented):
I'll apply remaining white spaces added by https://github.com/appium/appium-ios-device/pull/202/commits/f30676207a049047e7088a68ac8ecc98c8104314

KazuCocoa commented:
fyi: appium xcuitest driver 9.10.3+ has this change

KazuCocoa commented:
Hi @erdncyz, congrats: the Appium project wants to compensate you for this contribution! Please reply to this comment mentioning @jlipps and @KazuCocoa and sharing your [OpenCollective](https://opencollective.com/) account name, so that we can initiate payment! Or let us know if you decline to receive compensation via OpenCollective. Thank you!

KazuCocoa commented:
(Responded on slack)
